### PR TITLE
smartConfig support

### PR DESCRIPTION
**Pull request by seopyoon**
I added smart Config support for the library.

Based upon ESP83266 arduino library.

Haven't had chance to confirm it working.

### CODE REVIEW

File: libraries/WiFi/src/WiFiSTA.cpp [hunk 1]
```diff
@@ -498,3 +498,62 @@ IPv6Address WiFiSTAClass::localIPv6()
...
+
+    return false;
+}
+
+bool WiFiSTAClass::smartConfigDone() {
+    if (!_smartConfigStarted) {
+        return false;
+    }
+
+    return _smartConfigDone;
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiSTA.cpp]: i do not see _smartConfigDone being set to true anywhere? Maybe missing from the callback?

File: libraries/WiFi/src/WiFiSTA.cpp [hunk 2]
```diff
@@ -498,3 +498,62 @@ IPv6Address WiFiSTAClass::localIPv6()
     }
     return IPv6Address(addr.addr);
 }
+
+
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiSTA.cpp]: do not forget to add ```#include <esp_smartconfig.h>``` up on top

**Comment by me-no-dev** [comment]
Please see my comments, with those changes I have tested it and its working :)

**Comment by me-no-dev** [comment]
You might want to add an example as well :)

**Comment by seopyoon** [comment]
Done that!

@me-no-dev how did you test it? my example code doesn't seem to work.  please verify.  Thanks in advance.

**Comment by me-no-dev** [comment]
@seopyoon you did not add ```_smartConfigDone = true``` inside ```_smartConfigCallback```, so you never get the done status. Yes with the edits I mentioned I tested the code and it worked fine

**Comment by seopyoon** [comment]
Done!  I should've read your comments more thoroughly.  Could you confirm my example .ini working?

Thanks!

**Comment by me-no-dev** [comment]
Sketch filename should be the same as the folder name it's in. So change SmartConfig.ino to WiFiSmartConfig.ino
Also this sketch I think looks better and more informative on what is going on:
```arduino
#include "WiFi.h"

void setup() {
  Serial.begin(115200);

  //Init WiFi and SmartConfig
  WiFi.mode(WIFI_AP_STA);
  WiFi.beginSmartConfig();

  //Wait for SmartConfig packet from your phone
  Serial.print("Waiting for SmartConfig");
  while(!WiFi.smartConfigDone()){
    delay(500);
    Serial.print(".");
  }
  Serial.println("");
  Serial.println("SmartConfig recieved");

  //Wait for WiFi to connect with the recieved settings
  Serial.print("Waiting for WiFi");
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  Serial.println("");

  //WiFi is connected and we have IP
  Serial.println("WiFi Connected.");
  Serial.print("IP Address: ");
  Serial.println(WiFi.localIP());
}

void loop() {

}

```

Please do test your code and at least compile it.

**Comment by seopyoon** [comment]
@me-no-dev modified name of file, added comments.  Confirmed compiling, and running without error.  But, I cannot make it connect via ESP Touch technology.  I am using this [app](https://espressif.com/en/support/download/apks?keys=&field_technology_tid%5B%5D=20) for Android.  
Any idea?

**Comment by me-no-dev** [comment]
I also use the ESPTouch app from the App store (iPhone user here) and it was working. maybe check the android store for updated version?

**Comment by me-no-dev** [comment]
```
rst:0x1 (POWERON_RESET),boot:0x16 (SPI_FAST_FLASH_BOOT)
ets Jun  8 2016 00:22:57

rst:0x10 (RTCWDT_RTC_RESET),boot:0x16 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0x00
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0008,len:8
load:0x3fff0010,len:1760
load:0x40078000,len:6668
load:0x40080000,len:252
entry 0x40080034
Waiting for SmartConfig......................................................
SmartConfig received.
Waiting for WiFi......
WiFi Connected.
IP Address: 192.168.254.199
```

**Comment by christophepersoz** [comment]
Hello, 
I just tried the SmartConfig.ino in association with ESP touch for iOS and I can't get it connected. On the ESP32 side the sketch build and run normally, but on the iOS side nothing occurs once "Confirm" button pressed. Have you notice any issues across the versions that can affect the behaviours of the smartConfig support as the last related comment last from Jan 2017? 
Thanks.

**Pull request merged**